File: algorithm/counting_bits/go/counting_bits.go
package main

import (
	"fmt"
)

func countBits(num int) []int {
	result := make([]int, num+1)
	if num == 0 {
		return result
	}
	for i := 1; i <= num; i++ {
		result[i] = result[i&(i-1)] + 1
	}
	return result
}

func main() {
	r := countBits(2)
	fmt.Println(r)
}

File: algorithm/counting_bits/counting_bits.cpp
#include <iostream>
#include <vector>

using std::vector;
using std::cout;
using std::endl;

class Solution {
public:
	vector<int> countBits(int num) {
		vector<int> r(num+1, 0);
		if (num == 0) {
			return r;
		}
		for (int i = 1; i <= num; ++i) {
			r[i] = r[i&(i-1)]+1;
		}
		return r;

	}
};

int main(void) {
	Solution s;
	vector<int> result = s.countBits(123456);
	for (auto ite:result) {
		std::cout<<ite<<" ";
	}
	cout << endl;
	return 0;
}

